Deprecate torch.cross default behaviour

lezcano opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #108760

Long overdue this one. We may be able to change it in a few years :hopeful:.

**BC-breaking note**

This PR deprecates `torch.cross`'s default dim in favor of
`torch.linalg.cross`.
A upgrade guide is added to the documentation for `torch.cross`.

Note this PR DOES NOT remove `torch.cross`.

Fixes https://github.com/pytorch/pytorch/issues/108664

cc @jianyuh @nikitaved @pearu @mruberry @walterddr @IvanYashchuk @xwang233 @Lezcano

lezcano committed:
Deprecate torch.cross default behaviour

Long overdue this one. We may be able to change it in a few years :hopeful:.

**BC-breaking note**

This PR deprecates `torch.cross`'s default dim in favor of
`torch.linalg.cross`.
A upgrade guide is added to the documentation for `torch.cross`.

Note this PR DOES NOT remove `torch.cross`.

Fixes https://github.com/pytorch/pytorch/issues/108664

[ghstack-poisoned]

vadimkantorov commented:
@lezcano Please also remove default-dim from softmax :)
- https://github.com/pytorch/pytorch/issues/84524
- https://github.com/pytorch/pytorch/issues/36524

lezcano commented:
That Softmax deprecation may be more painful, and it'll need to be done by someone at meta. Feel free to poke them in that other issue.

albanD commented on aten/src/ATen/native/Cross.cpp:
```diff
@@ -59,6 +59,13 @@ static int64_t _default_cross_dim(const c10::optional<int64_t> &dimension, SymIn
+  if (!dimension) {
+    TORCH_WARN_ONCE(
+      "Using torch.cross without specifying the dim arg is deprecated.\n",
+      "Please either pass the dim explicitly or simply use linalg.cross.\n",
```
nit:
```suggestion
      "Please either pass the dim explicitly or simply use torch.linalg.cross.\n",
```

albanD reviewed (approved):
nit on the doc but sounds good otherwise

lezcano committed:
Update on "Deprecate torch.cross default behaviour"


Long overdue this one. We may be able to change it in a few years :hopeful:.

**BC-breaking note**

This PR deprecates `torch.cross`'s default dim in favor of
`torch.linalg.cross`.
A upgrade guide is added to the documentation for `torch.cross`.

Note this PR DOES NOT remove `torch.cross`.

Fixes https://github.com/pytorch/pytorch/issues/108664

cc jianyuh nikitaved pearu mruberry walterddr IvanYashchuk xwang233 Lezcano

[ghstack-poisoned]

lezcano commented:
@pytorchbot merge
